src/nbl/video/ILogicalDevice.cpp

```diff
@@ -34,50 +34,34 @@ bool ILogicalDevice::updateDescriptorSets(uint32_t descriptorWriteCount, const I
-        auto* samplers = ds->getMutableSamplers(pDescriptorWrites[i].binding);
-        for (auto j = 0; j < pDescriptorWrites[i].count; ++j)
-        {
-            descriptors[j] = pDescriptorWrites[i].info[j].desc;
 
-            if (samplers)
-                samplers[j] = pDescriptorWrites[i].info[j].info.image.sampler;
-        }
+        ds->incrementVersion();
+        ds->processWrite(pDescriptorWrites[i]);
     }
 
     for (auto i = 0; i < descriptorCopyCount; ++i)
     {
         const auto* srcDS = static_cast<const IGPUDescriptorSet*>(pDescriptorCopies[i].srcSet);
         auto* dstDS = static_cast<IGPUDescriptorSet*>(pDescriptorCopies[i].dstSet);
 
-        auto foundBindingInfo = std::lower_bound(srcDS->getLayout()->getBindings().begin(), srcDS->getLayout()->getBindings().end(), pDescriptorCopies[i].srcBinding,
-            [](const IGPUDescriptorSetLayout::SBinding& a, const uint32_t b) -> bool
-            {
-                return a.binding < b;
-            });
-
-        if (foundBindingInfo->binding != pDescriptorCopies[i].srcBinding)
-            return false;
+        // dstDS->processCopy(pDescriptorCopies[i]);
 
-        const asset::E_DESCRIPTOR_TYPE descriptorType = foundBindingInfo->type;
+        for (uint32_t t = 0; t < static_cast<uint32_t>(asset::IDescriptor::E_TYPE::ET_COUNT); ++t)
+        {
+            const auto type = static_cast<asset::IDescriptor::E_TYPE>(t);
 
-        auto* srcDescriptors = srcDS->getDescriptors(descriptorType, pDescriptorCopies[i].srcBinding);
-        auto* srcSamplers = srcDS->getMutableSamplers(pDescriptorCopies[i].srcBinding);
-        if (!srcDescriptors)
-            return false;
+            auto* srcDescriptors = srcDS->getDescriptors(type, pDescriptorCopies[i].srcBinding);
+            auto* srcSamplers = srcDS->getMutableSamplers(pDescriptorCopies[i].srcBinding);
 
-        auto* dstDescriptors = dstDS->getDescriptors(descriptorType, pDescriptorCopies[i].dstBinding);
-        auto* dstSamplers = dstDS->getMutableSamplers(pDescriptorCopies[i].dstBinding);
-        if (!dstDescriptors)
-            return false;
+            auto* dstDescriptors = dstDS->getDescriptors(type, pDescriptorCopies[i].dstBinding);
+            auto* dstSamplers = dstDS->getMutableSamplers(pDescriptorCopies[i].dstBinding);
 
-        // TODO(achal): Use copy_n.
-        memcpy(dstDescriptors, srcDescriptors, pDescriptorCopies[i].count * sizeof(core::smart_refctd_ptr<const asset::IDescriptor>));
+            if (srcDescriptors && dstDescriptors)
+                std::copy_n(srcDescriptors, pDescriptorCopies[i].count, dstDescriptors);
 
-        if (srcSamplers && dstSamplers)
-            memcpy(dstSamplers, srcSamplers, pDescriptorCopies[i].count * sizeof(core::smart_refctd_ptr<const IGPUSampler>));
+            if (srcSamplers && dstSamplers)
+                std::copy_n(srcSamplers, pDescriptorCopies[i].count, dstSamplers);
```

Comment:
you forgot to call `incrementVersion()`

Reply:
or actually lets fold it into `processCopy` and `processWrite`

Reply:
> you forgot to call `incrementVersion()`

in the copy loop

Reply:
> or actually lets fold it into `processCopy` and `processWrite`

If we do this then the descriptor set version will be incremented for each binding! If a descriptor set has 4 bindings, after updating it its version will become +4 of the original.

Reply:
Technically you could update a DS with a combination of writes and copies so `incrementVersion` should've been only called once. Yes, I know, I had it in the loop before as well which was wrong.

Reply:
Okay this is slightly more tricky because I don't think there is a VU which prohibits different `dstSet` members in `pDescriptorWrites` and `pDescriptorCopies`. In other words, I think its valid to update multiple DS at once. Therefore sometimes we have to `incrementVersion` and sometimes not (depending upon if we hit a new DS).

Reply:
well, there's no way around it, also the counter is 64bit for this reason :D

64bit counter means 2^64-1 increments before an overflow, assuming that the CPU is 5GHz and can do `processWrite` or `processCopy` in about 50 cycles (which is very generous, its probably more):
- 100 Million Updates/second
- 51'240'956 hours till overflow = 5845 years till overflow

Reply:
Another problem with this will arise when one of the either (write or copy) fails because then we increment the version of the sets while `updateDescriptorSets` fail.

Reply:
> Another problem with this will arise when one of the two (write or copy) fail because then we increment the version of the sets while `updateDescriptorSets` fail.

please explain?